cache/sharded_cache.cc

```diff
@@ -159,4 +159,25 @@ int GetDefaultCacheShardBits(size_t capacity) {
   return num_shard_bits;
 }
 
+namespace {
+// To access protected Status API
+class InsertStatusOkOverwritten : public Status {
+ public:
+  InsertStatusOkOverwritten()
+      : Status(Status::kOk, static_cast<Status::SubCode>(1)) {
+    // Confirm compatible with normal OK checks
+    assert(ok());
+    assert(*this == Status::OK());
+  }
+};
+}  // anonymous namespace
+
+const Status ShardedCache::kInsertStatusOkOverwritten =
+    InsertStatusOkOverwritten();
+
+bool ShardedCache::IsInsertStatusOkOverwritten(const Status& s) {
+  return s.code() == kInsertStatusOkOverwritten.code() &&
+         s.subcode() == kInsertStatusOkOverwritten.subcode();
+}
+
```

Comment:
I don't feel like this extension to Status is seasoned enough to bake into the public API (forever). In fact, one could argue that using variants of OK status for logic is an anti-pattern. We're doing it here because (a) we're wedging something into an existing public API, and (b) only using it for stats gathering.

So I think leaving the Status public API as it is the the lesser of evils for now.

Reply:
If that's an anti-pattern, I would argue this is a bigger one. :) It introduces the same subtype of OK and breaks the `code_ == kOk => subcode_ == kNone` invariant but does it in a sneakier way (it's not obvious to someone reading the code of `Status` that the invariant no longer holds). It also hijacks the pre-existing enum value 1 in the process, so now `kInsertStatusOkOverwritten.subcode() == kMutexTimeout`.

From a C++ perspective, there are a couple of issues (granted, these are somewhat theoretical; I'm not saying the current code would blow up in practice):
* `Status` is not meant to be a base class. It does not have a virtual destructor, so technically speaking `Status* status = new InsertStatusOkOverwritten; delete status;` is undefined behavior.
* Assignments like `const Status ShardedCache::kInsertStatusOkOverwritten = InsertStatusOkOverwritten();` are susceptible to the slicing problem. (If `kInsertStatusOkOverwritten` had members of its own, they would be sliced off here.)

Reply:
One more thing regarding the public API: I feel that the `Cache` interface is special in the sense that its intended client is the RocksDB library itself, not the application. So as long as we change the API and the RocksDB code that uses it in lockstep, we're good. At least that's the theory. Obviously we can't stop people from taking `Cache` and using it for their own nefarious purposes. :)

Reply:
> If that's an anti-pattern, I would argue this is a bigger one. :)

I still disagree. Any change to public APIs puts a 1000x multiplier on the impact of any hygienic evil involved. Status is already a little confusing in its == semantics. Elaborating Ok Status with variants would extend the confusion.

My hack is extremely localized. You haven't convinced me otherwise. In fact, you called potential direct users of Cache "nefarious".

> it's not obvious to someone reading the code of Status that the invariant no longer holds

I'm just not worried about someone relying on that invariant (very unlikely, odd, and not future-proof) and also being a direct user of Cache::Lookup.

> I feel that the Cache interface is special in the sense that its intended client is the RocksDB library itself

Exposing a shared_ptr in BlockBasedTableOptions is pretty inviting to custom implementations, IMHO. If you think we should codify your intention, we can move all functions from Cache to a new, non-public CacheImpl (extending Cache) and use casts (like implementations of Cache::Handle). I'm not opposed to that if we don't expect an uprising.

I'm not against breaking the interface, but I would like to do that in big steps rather than little steps. This API change is a little step that's possible to avoid with a hack. The hack would obviously be removed when we make a "big step" breaking change to the interface. (I would not be surprised if we change/extend the interface to support optimizations.)

Reply:
> > If that's an anti-pattern, I would argue this is a bigger one. :)
> 
> I still disagree. Any change to public APIs puts a 1000x multiplier on the impact of any hygienic evil involved. Status is already a little confusing in its == semantics. Elaborating Ok Status with variants would extend the confusion.

Many codes already have subcodes; I don't see why OK can't have one. In fact, your patch also introduces an OK variant, albeit one that's pretending to be a "Timeout Acquiring Mutex". There is no difference in terms of backwards compatibility either: 1) both are safe except for explicit subcode checks, and 2) the whole issue of backwards compatibility is kind of moot since we control the (intended) client classes (e.g. `BlockBasedTable`) as well. TBH I don't see a difference between the two, other than your current one attempts to hide the new subcode and is more hackish.

> > I feel that the Cache interface is special in the sense that its intended client is the RocksDB library itself
> 
> Exposing a shared_ptr in BlockBasedTableOptions is pretty inviting to custom implementations, IMHO. If you think we should codify your intention, we can move all functions from Cache to a new, non-public CacheImpl (extending Cache) and use casts (like implementations of Cache::Handle). I'm not opposed to that if we don't expect an uprising.
> 
> I'm not against breaking the interface, but I would like to do that in big steps rather than little steps. This API change is a little step that's possible to avoid with a hack. The hack would obviously be removed when we make a "big step" breaking change to the interface. (I would not be surprised if we change/extend the interface to support optimizations.)

OK now I feel there's some confusion here. Yes, the `Cache` interface is public so people can provide their own cache implementations. What I meant is that the intended client (or caller) of any cache implementation (LRU, clock, or a fancy custom cache) is `BlockBasedTable` et al., which are under our control. Having said that, I'm not suggesting any change that would break the compilation for anybody, not even for folks who are using `Cache` for other purposes in their applications. `Cache::Insert` would still have the exact same signature and return `Status`; the only change is that `Status` that could take on an additional code+subcode combination.

Reply:
P.S. If someone has a custom `Cache` implementation, they would have to tweak it so that `Insert` returns the new `Status` upon overwrite, or they don't get the new stats. Again, this is true regardless of how the new code+subcode combination is added.